src/security/ddwaf_memres.h

```diff
@@ -96,4 +96,47 @@ class DdwafMemres {
   std::size_t strings_stored_{0};
 };
 
+/*
+ * A pool of ddwaf object vectors of sizes in powers in two. Used when the
+ * sizes of arrays or maps are not known in advance.
+ */
+template <typename DdwafObjType>
+requires std::is_base_of_v<ddwaf_object, DdwafObjType>
+class DdwafObjArrPool {
+ public:
+  DdwafObjArrPool(DdwafMemres &memres) : memres_{memres} {}
```

Comment:
the memory returned by `DdwafObjArrPool` is owned `DdwafMemres` (which is owned by the `Context`); since the pool is used to build data submitted to the waf, it needs to be live until the waf context is destroyed.

Note that `DdwafObjArrPool` is destroyed by the time `parse_json` / `parse_multipart` return; it's merely a helper to use the `DdwafMemres` more efficiently.